Step2 리뷰 부탁드립니다.

happy-dev-travel opened the pull request:
#updated
리뷰 감사합니다! Setter들을 제거하다 보니 깜짝 놀랐네요. id도 막 바꿔버릴 수 있었더라고요 앞으로 주의해야겠습니다 :)


----
연관관계 맵핑 후 테스트 유지보수 했습니다.

그리고 Question과 User사이에 Cascade를 설정해서 학습테스트를 추가로 진행해 보았습니다.
해보니 @ManyToOne쪽에서는 CaseCasde 에 Remove를 넣으면 안 되겠더라고요
혹시 양방향 관계를 설정할 때 많이 사용하는 Cascade 설정이 있을까요?

happy-dev-travel committed:
feat(domain): 연관관계 맵핑

happy-dev-travel committed:
refactor(answerRepository): 연관관계 맵핑으로 인한 테스트 실패 수정

happy-dev-travel committed:
refactor(questionRepository): 연관관계 맵핑으로 인한 테스트 실패 수정

happy-dev-travel committed:
refactor(test): test용 변수를 가진 base 클래스 추가

happy-dev-travel committed:
refactor(test): service 테스트 실패 수정

happy-dev-travel committed:
feat(test): 연관관계 테스트 추가

happy-dev-travel committed:
feat(test): (학습테스트)
Question과 User사이에 Cascade 모두 ALL로 해봄

limwoobin commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -15,8 +15,13 @@ public class Answer extends BaseEntity{
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "question_id", nullable = false)
+    private Question question;
```
연관관계를 잘 설정해주셨습니다 ㅎㅎ
혹시 fetch type을 LAZY로 설정해놓은 이유가 있으신가요??

happy-dev-travel replied:
김영한님 강의를 들었었는데요ㅎㅎ 기본적으로 성능을 위해 LAZY 설정을 해두고 사용하다가 비즈니스 필요에 따라 항상 필요하면 EAGER로 바꾸라고 하셨어서 일당 LAZY로 설정 했습니다

limwoobin replied:
ㅎㅎ좋은 방법이네요 
저 역시도 실무에서 동길님과 같은방법으로 사용하고있습니다!

limwoobin commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -61,20 +66,20 @@ public void setId(Long id) {
 
-    public void setWriterId(Long writerId) {
-        this.writerId = writerId;
+    public void setWriter(User writer) {
+        this.writer = writer;
     }
```
기존에 있는 메서드였지만 Entity내에서의 setter 메소드는 지양하는 편이 좋습니다 ㅎㅎ
객체지향의 성격을 떨어트릴수도 있습니다.
전체적으로 한번 제거하는것을 고려해주세요!

https://velog.io/@aidenshin/%EB%82%B4%EA%B0%80-%EC%83%9D%EA%B0%81%ED%95%98%EB%8A%94-JPA-%EC%97%94%ED%8B%B0%ED%8B%B0-%EC%9E%91%EC%84%B1-%EC%9B%90%EC%B9%99

happy-dev-travel replied:
기본 setter는 제거하고 의도가 드러나는 메소드를 써야 하는군요! 감사합니다

limwoobin replied:
필수는 아니지만 많이 권장하는 방법입니다!

limwoobin commented on src/main/java/qna/domain/User.java:
```diff
@@ -20,6 +22,12 @@ public class User {
     private String name;
     private String email;
 
+    @OneToMany(mappedBy = "writer", fetch = FetchType.LAZY, cascade = CascadeType.ALL)
+    private Set<Question> questions = new HashSet<>();
+
+    public void addQuestion(Question question){
+        this.questions.add(question);
+    }
```
혹시 User, Question 을 양방향으로 설정해주신 이유가 있을까요??

happy-dev-travel replied:
사실 Cascade를 처음 써봐서 어떻게 동작하나 궁금해서 양방향에 Casacade.ALL로 해놓고 테스트를 했습니다 :)

limwoobin replied:
그러시군요 이번 요구사항에서 양방향 연관관계가 필수로 있어야할지에 대해
고민해보시면 좋겠습니다 ㅎㅎ

happy-dev-travel replied:
아아 요구사항으로 보면 당연히 필수는 아닙니다 제거하고 올리겠습니다 :)

limwoobin commented on src/test/java/qna/domain/NewEntityTestBase.java:
```diff
@@ -0,0 +1,20 @@
+        Q1 = new Question("title", "contents").writeBy(NEWUSER1);
+        Q2 = new Question("title", "contents").writeBy(NEWUSER2);
+    }
+}
```
테스트에 필요한 데이터들을 이런방식으로 세팅해주셨군요 ㅎㅎ

limwoobin commented on src/test/java/qna/domain/UserRepositoryTest.java:
```diff
@@ -14,33 +14,31 @@
 
 @DataJpaTest
-class UserRepositoryTest {
+class UserRepositoryTest extends NewEntityTestBase{
```
자바 컨벤션 설정을 다시 확인해주세요!

happy-dev-travel replied:
컨벤션 중에 어떤 부분을 봐야하는 건지요?

limwoobin replied:
`NewEntityTestBase{` 에 공백이 없어서 말씀드렷어요

happy-dev-travel replied:
매의눈 이시군요 감사합니다 :)

limwoobin commented on src/main/java/qna/domain/User.java:
```diff
@@ -3,7 +3,9 @@
+import java.util.Set;
 
 @Entity
 public class User {
```
User의 미사용 getter / setter 메소드를 제거해주셨군요! ㅎㅎ

limwoobin reviewed (changes requested):
안녕하세요 동길님 2단계 미션 잘 진행해주셨습니다 ㅎㅎ
`@DataJpaTest` 이외에도 엔티티에 대한 단위테스트까지 
굉장히 꼼꼼하게 잘 작성해주셧네요~

간단한 코멘트 몇개 남겨두었으니 확인한번 부탁드립니다

limwoobin commented:
`User.java` 에 unique 설정이 빠진것같은데 확인부탁드립니다~

limwoobin commented:
> 연관관계 맵핑 후 테스트 유지보수 했습니다.
> 
> 그리고 Question과 User사이에 Cascade를 설정해서 학습테스트를 추가로 진행해 보았습니다. 해보니 @manytoone쪽에서는 CaseCasde 에 Remove를 넣으면 안 되겠더라고요 혹시 양방향 관계를 설정할 때 많이 사용하는 Cascade 설정이 있을까요?

말씀하신것처럼 @ManyToOne 에서 cascade 설정을 넣게되면 다른 관계를 맺고있는 데이터들에게도 영향이 갈 수 있습니다 👍 
보통 저는 양방향관계를 설정할때는 라이프사이클이 동일하거나 강한 의존을 맺고있는경우에 양방향을 맺고는 합니다
내부적인 cascade설정은 비즈니스에 따라 다르겠지만 일반적으로 라이프사이클을 기준으로 설정하고는 합니다 ㅎㅎ

happy-dev-travel committed:
refactor(test): 리뷰 반영

- User Entity의 userId에 Unique설정
- 자바 컨벤션 확인
- 불필요한 양방향 연관 관계 제거, 관련 테스트 제거

happy-dev-travel committed:
refactor(test): 리뷰 반영 2

- Entity의 Setter 함수들을 전반적으로 제거함.
- ManyToOne에 설정되어 있던 CascadeType.ALL을 Persist로 변경

limwoobin commented on src/main/java/qna/domain/Question.java:
```diff
@@ -8,11 +8,14 @@ public class Question extends BaseEntity {
     private String contents;
-    private Long writerId;
+
+    @ManyToOne(fetch = FetchType.LAZY, cascade = CascadeType.PERSIST)
```
cascade가 persist로 필요한 이유가 따로 있을까요??

limwoobin reviewed (approved):
동길님 전체적으로 피드백 잘 반영해주셨네요!
고생많으셨습니다.
코멘트하나 남겨드렸으니 다음미션하면서 같이 확인해주세요
다음미션도 화이팅입니다!
